File: scripts/fixed_width_to_csv.py
# /// script
# dependencies = [
#   "duckdb",
# ]
# ///

"""
Convert fixed-width space-separated text files to pipe-delimited CSV files.
"""

import subprocess
import sys
import os
from pathlib import Path

import duckdb

# From readme.txt of the GHCND data set:
# IV. FORMAT OF "ghcnd-stations.txt"

# ------------------------------
# Variable   Columns   Type
# ------------------------------
# ID            1-11   Character
# LATITUDE     13-20   Real
# LONGITUDE    22-30   Real
# ELEVATION    32-37   Real
# STATE        39-40   Character
# NAME         42-71   Character
# GSN FLAG     73-75   Character
# HCN/CRN FLAG 77-79   Character
# WMO ID       81-85   Character
# ------------------------------

# Convert the above list into a list of column names and start positions but not the end positions, one tuple each
column_names_and_start_positions = {'stations': [
    ('ID', 1),
    ('LATITUDE', 13),
    ('LONGITUDE', 22),
    ('ELEVATION', 32),
    ('STATE', 39),
    ('NAME', 42),
    ('GSN FLAG', 73),
    ('HCN/CRN FLAG', 77),
    ('WMO ID', 81),
]}


def extract_columns(line, column_starts):
    """
    Extract column values from a line using the identified column positions.
    
    Args:
        line: String (single line from file)
        column_starts: List of column start positions
        
    Returns:
        List of column values (stripped of leading/trailing spaces)
    """
    line = line.rstrip('\n')
    columns = []
    
    for i in range(len(column_starts)):
        if i < len(column_starts) - 1:
            # Extract from current start to next start
            col_value = line[column_starts[i]:column_starts[i+1]]
        else:
            # Last column: extract from start to end of line
            col_value = line[column_starts[i]:]
        
        # Strip trailing spaces from column value
        columns.append(col_value.rstrip())
    
    return columns


def convert_fixed_width_to_csv(input_filepath, output_filepath, data_set_name):
    """
    Convert a fixed-width file to a pipe-delimited CSV file.
    
    Args:
        input_filepath: Path to the input fixed-width file
    """
    # Read all lines from the input file
    try:
        with open(input_filepath, 'r', encoding='utf-8') as f:
            lines = f.readlines()
    except FileNotFoundError:
        print(f"Error: File '{input_filepath}' not found.")
        return
    except Exception as e:
        print(f"Error reading file: {e}")
        return
    
    if not lines:
        print("Error: Input file is empty.")
        return
    
    column_starts = [start_pos - 1 for _, start_pos in column_names_and_start_positions[data_set_name]]

    # Write the header row, but lowercase the column names and replace spaces with underscores
    header_row = [name.lower().replace(' ', '_') for name, _ in column_names_and_start_positions[data_set_name]]
    with open(output_filepath, 'w', encoding='utf-8') as f:
        f.write('|'.join(header_row) + '\n')
    
    # Convert and write to CSV
    try:
        with open(output_filepath, 'w', encoding='utf-8') as f:
            for line_num, line in enumerate(lines):
                # Extract columns from this line
                columns = extract_columns(line, column_starts)
                
                # Write as pipe-delimited CSV
                f.write('|'.join(columns) + '\n')
                
                # Progress indicator for large files
                if (line_num + 1) % 1000 == 0:
                    print(f"Processed {line_num + 1} lines...")
        
        print(f"\nConversion complete!")
        print(f"Input:  {input_filepath}")
        print(f"Output: {output_filepath}")
        print(f"Total lines processed: {len(lines)}")
        
    except Exception as e:
        print(f"Error writing output file: {e}")
        return
    
def download_from_s3_if_not_exists(s3_path, local_path):
    """
    Download a file from S3 to a local path if it does not exist.
    """
    if not os.path.exists(local_path):
        print(f"File '{local_path}' not found. Downloading from S3...")
        subprocess.run(["aws", "s3", "cp", "--no-sign-request", s3_path, local_path])
        print(f"Downloaded file '{local_path}' from S3.")
    else:
        print(f"File '{local_path}' already exists. Skipping download.")

def convert_to_parquet(input_filepath, output_filepath):
    """
    Convert a CSV file to a Parquet file using DuckDB.
    """
    duckdb.sql(f"COPY (FROM '{input_filepath}') TO '{output_filepath}' (PARQUET_VERSION V2, COMPRESSION ZSTD)")

def main():
    """Main entry point for the script."""
    if len(sys.argv) != 2:
        print(f"Usage: python {sys.argv[0]} <data_set_name>")
        print("\nThis script converts a fixed-width space-separated file to a")
        print("pipe-delimited CSV file. The data set name is used to determine")
        print("the column names and start positions.")
        sys.exit(1)
    
    data_set_name = sys.argv[1]
    input_filepath = f"data/ghcnd-{data_set_name}.txt"
    output_filepath = f"data/ghcnd-{data_set_name}.csv"
    # If the parquet file doesn't exist, convert the CSV to parquet 
    parquet_filepath = f"data/ghcnd-{data_set_name}.parquet"
    if not os.path.exists(parquet_filepath):
        download_from_s3_if_not_exists(f"s3://noaa-ghcn-pds/ghcnd-{data_set_name}.txt", input_filepath)
        convert_fixed_width_to_csv(input_filepath, output_filepath, data_set_name)
        convert_to_parquet(output_filepath, parquet_filepath)
    else:
        print(f"Parquet file '{parquet_filepath}' already exists. Skipping conversion.")


if __name__ == "__main__":
    main()

File: scripts/fixed_width_to_csv.js
// fixed_width_to_csv.js
// Translation of fixed_width_to_csv.py to JavaScript (Node.js)
// Dependencies: aws-cli (for S3), duckdb (Node.js package), fs, path

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const duckdb = require('duckdb');

// Column definitions as in the Python script
const columnNamesAndStartPositions = {
  stations: [
    ['ID', 1],
    ['LATITUDE', 13],
    ['LONGITUDE', 22],
    ['ELEVATION', 32],
    ['STATE', 39],
    ['NAME', 42],
    ['GSN FLAG', 73],
    ['HCN/CRN FLAG', 77],
    ['WMO ID', 81],
  ],
  countries: [
    ['CODE', 1],
    ['NAME', 4],
  ],
  states: [
    ['CODE', 1],
    ['NAME', 4],
  ],
};

function extractColumns(line, columnStarts) {
  const columns = [];
  for (let i = 0; i < columnStarts.length; i++) {
    const start = columnStarts[i];
    const end = i < columnStarts.length - 1 ? columnStarts[i + 1] : undefined;
    const colValue = end ? line.slice(start, end) : line.slice(start);
    columns.push(colValue.trimEnd());
  }
  return columns;
}

function convertFixedWidthToCsv(inputFilePath, outputFilePath, dataSetName) {
  let lines;
  try {
    lines = fs.readFileSync(inputFilePath, 'utf-8').split('\n');
  } catch (err) {
    console.error(`Error: File '${inputFilePath}' not found or unreadable.`);
    return;
  }
  if (!lines.length || (lines.length === 1 && lines[0] === '')) {
    console.error('Error: Input file is empty.');
    return;
  }
  const columnStarts = columnNamesAndStartPositions[dataSetName].map(([, start]) => start - 1);
  const headerRow = columnNamesAndStartPositions[dataSetName].map(([name]) => name.toLowerCase().replace(/ /g, '_'));
  const out = fs.createWriteStream(outputFilePath, { encoding: 'utf-8' });
  out.write(headerRow.join('|') + '\n');
  lines.forEach((line, idx) => {
    if (!line.trim()) return;
    const columns = extractColumns(line, columnStarts);
    out.write(columns.join('|') + '\n');
    if ((idx + 1) % 1000 === 0) {
      console.log(`Processed ${idx + 1} lines...`);
    }
  });
  out.end();
  console.log(`\nConversion complete!`);
  console.log(`Input:  ${inputFilePath}`);
  console.log(`Output: ${outputFilePath}`);
  console.log(`Total lines processed: ${lines.length}`);
}

function downloadFromS3IfNotExists(s3Path, localPath) {
  if (!fs.existsSync(localPath)) {
    console.log(`File '${localPath}' not found. Downloading from S3...`);
    execSync(`aws s3 cp --no-sign-request ${s3Path} ${localPath}`, { stdio: 'inherit' });
    console.log(`Downloaded file '${localPath}' from S3.`);
  } else {
    console.log(`File '${localPath}' already exists. Skipping download.`);
  }
}

function convertToParquet(inputFilePath, outputFilePath) {
  // Using DuckDB Node.js API
  console.log(`Converting ${inputFilePath} to ${outputFilePath}`);
  const db = new duckdb.Database(':memory:');
  db.run(
    `COPY (FROM read_csv('${inputFilePath}', delim='|', header=true)) TO '${outputFilePath}' (PARQUET_VERSION V2, COMPRESSION ZSTD)`,
    (err) => {
      if (err) {
        console.error('DuckDB error:', err);
      } else {
        console.log(`Converted ${inputFilePath} to ${outputFilePath}`);
      }
      db.close();
    }
  );
}

function convertDataSet(dataSetName) {
  const inputFilePath = `data/ghcnd-${dataSetName}.txt`;
  const outputFilePath = `data/ghcnd-${dataSetName}.csv`;
  const parquetFilePath = `data/ghcnd-${dataSetName}.parquet`;
  if (!fs.existsSync(parquetFilePath)) {
    downloadFromS3IfNotExists(`s3://noaa-ghcn-pds/ghcnd-${dataSetName}.txt`, inputFilePath);
    convertFixedWidthToCsv(inputFilePath, outputFilePath, dataSetName);
    convertToParquet(outputFilePath, parquetFilePath);
  } else {
    console.log(`Parquet file '${parquetFilePath}' already exists. Skipping conversion.`);
  }
}

function main() {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    Object.keys(columnNamesAndStartPositions).sort().forEach(convertDataSet);
  } else if (args.length === 1) {
    convertDataSet(args[0]);
  } else {
    console.log(`Usage: node ${path.basename(process.argv[1])} [<data_set_name>]`);
    console.log('\nThis script converts a fixed-width space-separated file to a');
    console.log('pipe-delimited CSV file. The data set name is used to determine');
    console.log('the column names and start positions.');
    console.log('If no data set name is provided, all data sets will be processed.');
    process.exit(1);
  }
}

if (require.main === module) {
  main();
} 